presto-bigquery/src/main/java/com/facebook/presto/plugin/bigquery/BigQueryResultPageSource.java

```diff
@@ -0,0 +1,342 @@
+                .setReadPosition(Storage.StreamPosition.newBuilder()
+                        .setStream(Storage.Stream.newBuilder()
+                                .setName(split.getStreamName())));
+        responses = new ReadRowsHelper(bigQueryStorageClient, readRowsRequest, maxReadRowsRetries).readRows();
+        closed = false;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return readBytes.get();
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return !responses.hasNext();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        checkState(pageBuilder.isEmpty(), "PageBuilder is not empty at the beginning of a new page");
+        Storage.ReadRowsResponse response = responses.next();
+        Iterable<GenericRecord> records = parse(response);
+        for (GenericRecord record : records) {
+            pageBuilder.declarePosition();
+            for (int column = 0; column < columnTypes.size(); column++) {
+                BlockBuilder output = pageBuilder.getBlockBuilder(column);
+                appendTo(columnTypes.get(column), record.get(column), output);
+            }
+        }
+
+        Page page = pageBuilder.build();
+        pageBuilder.reset();
+        return page;
+    }
+
+    private void appendTo(Type type, Object value, BlockBuilder output)
+    {
+        if (value == null) {
+            output.appendNull();
+            return;
+        }
+
+        Class<?> javaType = type.getJavaType();
+        try {
+            if (javaType == boolean.class) {
+                type.writeBoolean(output, (Boolean) value);
+            }
+            else if (javaType == long.class) {
+                if (type.equals(BIGINT)) {
+                    type.writeLong(output, ((Number) value).longValue());
+                }
+                else if (type.equals(INTEGER)) {
+                    type.writeLong(output, ((Number) value).intValue());
+                }
+                else if (type.equals(DATE)) {
+                    type.writeLong(output, ((Number) value).intValue());
+                }
+                else if (type.equals(TIMESTAMP)) {
+                    type.writeLong(output, BigQueryType.toPrestoTimestamp(((org.apache.avro.util.Utf8) value).toString()));
+                }
+                else if (type.equals(TIME_WITH_TIME_ZONE)) {
+                    type.writeLong(output, DateTimeEncoding.packDateTimeWithZone(((Long) value).longValue() / 1000, TimeZoneKey.UTC_KEY));
+                }
+                else if (type.equals(TIMESTAMP_WITH_TIME_ZONE)) {
+                    type.writeLong(output, DateTimeEncoding.packDateTimeWithZone(((Long) value).longValue() / 1000, TimeZoneKey.UTC_KEY));
+                }
```

Comment:
Merge these two?  something like if (A || B)? or using a switch might help in case there might be more types need to be considered here.

Reply:
logical or op makes sense